### PR TITLE
[Rollup] Log deprecation if config and request intervals are mixed

### PR DESCRIPTION
**Pull request by polyfractal**
Backport of #32052, with some changes.

In 6.5, we forbid mixed intervals (one fixed, one calendar).  But for a 6.4.x bugfix release this would be a pretty extreme break.  So we continue to allow mixed intervals in 6.4, but log a deprecation warning.

6.4 also doesn't enforce that the request is a multiple of the config. So we log a deprecation if that is not the case, but don't prevent the query.

Finally, mixed intervals are only used if a better job cap can't be found.

To support the case where we have to compare fixed and calendar, there is a table of rough conversions (e.g. 1 day == 86400000ms).  This lets us crudely ensure the user doesn't try to query `2d` on an index that is set for `1w`.

@jimczi Mind giving a quick review on this?  The major change from the other PR is the addition of `validateMixedInterval()`, changes to the `CALENDAR_ORDERING` table, and that we don't enforce multiples.

### CODE REVIEW

File: x-pack/docs/en/rollup/rollup-search-limitations.asciidoc [hunk 1]
```diff
@@ -80,9 +80,25 @@ The response will tell you that the field and aggregation were not possible, bec
 [float]
 === Interval Granularity
 
-Rollups are stored at a certain granularity, as defined by the `date_histogram` group in the configuration.  If data is rolled up at hourly
-intervals, the <<rollup-search>> API can aggregate on any time interval hourly or greater.  Intervals that are less than an hour will throw
-an exception, since the data simply doesn't exist for finer granularities.
+Rollups are stored at a certain granularity, as defined by the `date_histogram` group in the configuration.  This means you
+can only search/aggregate the rollup data with an interval that is greater-than or equal to the configured rollup interval.
```
Comment (polyfractal) [inline x-pack/docs/en/rollup/rollup-search-limitations.asciidoc]: This isn't technically true, but I'm inclined to leave as-is to encourage proper config in 6.4 :)

File: x-pack/docs/en/rollup/rollup-search-limitations.asciidoc [hunk 2]
```diff
@@ -80,9 +80,25 @@ The response will tell you that the field and aggregation were not possible, bec
...
+.Requests must be multiples of the config
+**********************************
+Perhaps not immediately apparent, but the interval specified in an aggregation request must be a whole
+multiple of the configured interval.  If the job was configured to rollup on `3d` intervals, you can only
+query and aggregate on multiples of three (`3d`, `6d`, `9d`, etc).
+
+A non-multiple wouldn't work, since the rolled up data wouldn't cleanly "overlap" with the buckets generated
+by the aggregation, leading to incorrect results.
+
+For that reason, an error is thrown if a whole multiple of the configured interval isn't found.
```
Comment (polyfractal) [inline x-pack/docs/en/rollup/rollup-search-limitations.asciidoc]: Same as above.

**Comment by polyfractal** [comment]
Note: I also [spotted a bug in the comparator](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupJobIdentifierUtils.java#L347), similar to the parsing bug that's being fixed.  But I'd like to fix that in a followup PR because it applies to all branches.

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupJobIdentifierUtils.java [hunk 3]
```diff
@@ -133,6 +191,105 @@ private static void checkDateHisto(DateHistogramAggregationBuilder source, List<
...
+    static boolean validateCalendarInterval(DateHistogramInterval requestInterval, DateHistogramInterval configInterval) {
+        // Both must be calendar intervals
+        if (isCalendarInterval(requestInterval) == false || isCalendarInterval(configInterval) == false) {
+            return false;
+        }
+
+        // The request must be gte the config.  The CALENDAR_ORDERING map values are integers representing
+        // relative orders between the calendar units
+        long requestOrder = CALENDAR_ORDERING.getOrDefault(requestInterval.toString(), Long.MAX_VALUE);
+        long configOrder = CALENDAR_ORDERING.getOrDefault(configInterval.toString(), Long.MAX_VALUE);
```
Comment (jimczi) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupJobIdentifierUtils.java]: You can maybe replace the static map with:
````
DateTimeUnit configUnit = DateHistogramAggregationBuilder.DATE_FIELD_UNITS.get(configInterval.toString());
long configOrder = configUnit.field(DateTimeZone.UTC).getDurationField().getUnitMillis();
````
?

**Review by jimczi — COMMENTED** [review]
It looks good @polyfractal , I left some comments

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupJobIdentifierUtils.java (hunk 3 above)
Comment (polyfractal) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupJobIdentifierUtils.java]: Didn't realize you could do this.  ++

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupJobIdentifierUtils.java [hunk 4]
```diff
@@ -144,8 +301,8 @@ private static void checkHisto(HistogramAggregationBuilder source, List<RollupJo
                 for (Map<String, Object> agg : fieldCaps.getAggs()) {
                     if (agg.get(RollupField.AGG).equals(HistogramAggregationBuilder.NAME)) {
                         Long interval = (long)agg.get(RollupField.INTERVAL);
-                        // TODO should be divisor of interval
-                        if (interval <= source.interval()) {
+                        // query interval must be gte the configured interval, and a whole multiple
+                        if (interval <= source.interval() && source.interval() % interval == 0) {
```
[the remark this answers is not in the document]
Comment (polyfractal) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupJobIdentifierUtils.java]: argh, yes.  adding thanks

**Comment by polyfractal** [comment]
Tidied up, and added a few more tests for the multiples warning

**Pull request merged**